query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,486 @@
+                let values = children.into_iter()
+                                        .map(|arg| {
+                                            // We need to get conversion errors out.
+                                            let r = self.typed_value_from_arg(schema, arg, known_types.clone());
+                                            if let Ok(ref tv) = r {
+                                                if accumulated_types.insert(tv.value_type()) &&
+                                                   !accumulated_types.is_unit() {
+                                                    // Values not all of the same type.
+                                                    bail!(ErrorKind::InvalidGroundConstant);
+                                                }
+                                            }
+                                            r
+                                        })
+                                        .collect::<Result<Vec<_>>>()?;
+
+                // Otherwise, we now have the values and the type.
+                let types = vec![accumulated_types.exemplar().unwrap()];
+                let names = vec![var.clone()];
+
+                self.collect_named_bindings(schema, names, types, values);
+                Ok(())
+            },
+
+            (Binding::BindRel(places), FnArg::Vector(rows)) => {
+                if rows.is_empty() {
+                    bail!(ErrorKind::InvalidGroundConstant);
+                }
+
+                // Grab the known types to which these args must conform, and track
+                // the places that won't be bound in the output.
+                let template: Vec<(bool, ValueTypeSet)> =
+                    places.iter()
+                          .map(|x| match x {
+                              &VariableOrPlaceholder::Placeholder     => (false, ValueTypeSet::any()),
+                              &VariableOrPlaceholder::Variable(ref v) => (true, self.known_type_set(v)),
+                          })
+                          .collect();
+
+                // The expected 'width' of the matrix is the number of named variables.
+                let full_width = places.len();
+                let names: Vec<Variable> = places.into_iter().filter_map(|x| x.into_var()).collect();
+                let expected_width = names.len();
+                let expected_rows = rows.len();
+
+                if expected_width == 0 {
+                    // They can't all be placeholders.
+                    bail!(ErrorKind::InvalidGroundConstant);
+                }
+
+                // Accumulate values into `matrix` and types into `a_t_f_c`.
+                // This representation of a rectangular matrix is more efficient than one composed
+                // of N separate vectors.
+                let mut matrix = Vec::with_capacity(expected_width * expected_rows);
+                let mut accumulated_types_for_columns: Vec<ValueTypeSet> =
```

Comment:
There's a version of `vec!` that takes an element and a count; it should be both clearer and more efficient than this.